qupath-core/src/main/java/qupath/lib/objects/utils/ObjectMerger.java

```diff
@@ -189,15 +190,13 @@ private List<List<PathObject>> computeClustersIterative(Collection<? extends Pat
                     else
                         allPotentialNeighbors = allObjects;
                     var neighbors = filterCompatibleNeighbors(current, allPotentialNeighbors);
-                    for (var neighbor : neighbors) {
-                        if (!alreadyVisited.contains(neighbor)) {
-                            if (mergeTest.test(
+                    var addable = neighbors.stream()
+                            .filter(neighbor -> !alreadyVisited.contains(neighbor)) // alreadyVisited (set) is not thread-safe
+                            .parallel()
+                            .filter(neighbor -> mergeTest.test(
                                     currentGeometry,
-                                    getGeometry(neighbor, geometryMap))) {
-                                pending.add(neighbor);
-                            }
-                        }
-                    }
+                                    getGeometry(neighbor, geometryMap)));
```

Comment:
Here, the line `.filter(neighbor -> !alreadyVisited.contains(neighbor))` is still executed concurrently. For example, the code:

```java
var integers = IntStream.range(0, 10)
        .boxed()
        .peek(i -> System.err.println(Thread.currentThread()))
        .parallel()
        .toList();
System.err.println(integers);
```

prints:

```
Thread[main,5,main]
Thread[ForkJoinPool.commonPool-worker-2,5,main]
Thread[ForkJoinPool.commonPool-worker-5,5,main]
Thread[ForkJoinPool.commonPool-worker-1,5,main]
Thread[ForkJoinPool.commonPool-worker-7,5,main]
Thread[main,5,main]
Thread[ForkJoinPool.commonPool-worker-3,5,main]
Thread[ForkJoinPool.commonPool-worker-1,5,main]
Thread[ForkJoinPool.commonPool-worker-5,5,main]
Thread[ForkJoinPool.commonPool-worker-4,5,main]
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
```

This is because `parallel()` is an [intermediary operation](https://docs.oracle.com/javase/8/docs/api/java/util/stream/package-summary.html#StreamOps), so it makes the whole stream parallel (not only what happens after its call).

Reply:
Interesting! I guess this monstrosity overcomes that...?
```java
var integers = IntStream.range(0, 10)
        .boxed()
        .filter(i -> true)
        .toList()
        .stream()
        .parallel()
        .toList();
System.err.println(integers);
```

Reply:
Yes, because `toList()` is a terminal operation.

Reply:
That's unintuitive, though the justifications make sense. Should be fixed now

Reply:
The reason is explained in the [Javadocs](https://docs.oracle.com/javase/8/docs/api/java/util/stream/package-summary.html#StreamOps):

>  Processing streams lazily allows for significant efficiencies; in a pipeline such as the filter-map-sum example above, filtering, mapping, and summing can be fused into a single pass on the data, with minimal intermediate state. Laziness also allows avoiding examining all the data when it is not necessary; for operations such as "find the first string longer than 1000 characters", it is only necessary to examine just enough strings to find one that has the desired characteristics without examining all of the strings available from the source. (This behavior becomes even more important when the input stream is infinite and not merely large.)